pallets/runtime/tests/src/asset_metadata_test.rs

```diff
@@ -0,0 +1,427 @@
+            AssetError::AssetMetadataLocalKeyAlreadyExists
+        );
+    });
+}
+
+#[test]
+fn register_asset_metadata_global_type() {
+    ExtBuilder::default().build().execute_with(|| {
+        let alice = User::new(AccountKeyring::Alice);
+        let root = Origin::from(frame_system::RawOrigin::Root);
+
+        let (name, spec) = make_metadata_type("TEST");
+
+        // Try registering global metadata type with non-root.
+        exec_noop!(
+            Asset::register_asset_metadata_global_type(alice.origin(), name.clone(), spec.clone(),),
+            DispatchError::BadOrigin
+        );
+
+        // Register global metadata type with root.
+        assert_ok!(Asset::register_asset_metadata_global_type(
+            root.clone(),
+            name.clone(),
+            spec.clone(),
+        ));
+
+        // Try registering metadata with the same name.
+        assert_noop!(
+            Asset::register_asset_metadata_global_type(root, name, spec,),
+            AssetError::AssetMetadataGlobalKeyAlreadyExists
+        );
+    });
+}
+
+#[test]
+fn register_asset_metadata_local_type_limits() {
+    ExtBuilder::default().build().execute_with(|| {
+        let owner = User::new(AccountKeyring::Dave);
+
+        // Create asset.
+        let (ticker, _) = create_token(owner);
+
+        // Try registering metadata with over-sized values.
+        let register_type = |name, url, desc, type_def, err: DispatchError| {
+            let (name, spec) = make_metadata_type_sizes(name, url, desc, type_def);
+            exec_noop!(
+                Asset::register_asset_metadata_local_type(owner.origin(), ticker, name, spec,),
+                err
+            );
+        };
+
+        // Oversized metadata name.
+        register_type(
+            AssetMetadataNameMaxLength::get() + 1,
+            10,
+            10,
+            10,
+            AssetError::AssetMetadataNameMaxLengthExceeded.into(),
+        );
+
+        // Oversized metadata url.
+        register_type(10, MaxLen::get() + 1, 10, 10, BaseError::TooLong.into());
+
+        // Oversized metadata description.
+        register_type(10, 10, MaxLen::get() + 1, 10, BaseError::TooLong.into());
+
+        // Oversized metadata type definition.
+        register_type(
+            10,
+            10,
+            10,
+            AssetMetadataTypeDefMaxLength::get() + 1,
+            AssetError::AssetMetadataTypeDefMaxLengthExceeded.into(),
+        );
+    });
+}
+
+#[test]
+fn register_asset_metadata_global_type_limits() {
+    ExtBuilder::default().build().execute_with(|| {
+        let root = Origin::from(frame_system::RawOrigin::Root);
+
+        // Try registering metadata with over-sized values.
+        let register_type = |name, url, desc, type_def, err: DispatchError| {
+            let (name, spec) = make_metadata_type_sizes(name, url, desc, type_def);
+            assert_noop!(
+                Asset::register_asset_metadata_global_type(root.clone(), name, spec,),
```

Comment:
The test above looks sufficiently similar to this one that most of it could be de-duped with a higher order function.

Reply:
No.